src/Plugin/ocha_ai/VectorStore/ElasticsearchJob.php

```diff
@@ -45,69 +46,178 @@ public function getRelevantContents(string $index, array $ids, string $query_tex
     }
 
     $query = [
-      'size' => 50,
-      'query' => [
-        'nested' => [
-          'path' => 'contents',
-          'query' => [
-            'script_score' => [
-              'query' => [
-                'bool' => [
-                  'filter' => [
-                    'exists' => [
-                      'field' => 'contents.embedding',
-                    ],
-                  ],
-                ],
+      'knn' => [
+        'field' => 'embedding',
+        'query_vector' => $query_embedding,
+        'k' => (int) $this->getPluginSetting('topk'),
+        'num_candidates' => 250,
+      ],
+      '_source' => [
+        'excludes' => [
+          'embedding',
+        ],
+      ],
+    ];
+
+    try {
+      $response = $this->request('POST', $index . '/_knn_search', json_encode($query), 'application/json');
+      $data = json_decode($response->getBody()->getContents(), TRUE);
+
+      if (!is_null($data)) {
+        // Get the list of contents and their similarity score.
+        $contents = [];
+        foreach ($data['hits']['hits'] ?? [] as $hit) {
+          if (!in_array($hit['_id'], $ids)) {
+            $contents[$hit['_id']] = $hit['_score'];
+          }
+        }
+
+        if (empty($contents)) {
+          return [];
+        }
+
+        // Sort by score descending.
+        arsort($contents);
+        // Retrieve the minimum similarity to be considered relevant.
+        $cutoff = $this->getSimilarityScoreCutOff($contents, .20);
+
+        // Exclude irrelevant contents.
+        return array_filter($contents, function ($score) use ($cutoff) {
+          return $score >= $cutoff;
+        });
+      }
+    }
+    catch (GuzzleException $exception) {
```

Comment:
The catch added to the `::request()` method line 742 of `src/Plugin/ocha_ai/VectorStore/Elasticsearch.php` means that there won't be any GuzzleException thrown.

On the other hand, the `$response` form the `$this->request()` may be NULL in which PHP will complain line 64 when trying to get the body (`$response->getBody()`).

So I think this needs to be refactored a bit.